File: crabzilla-core/src/test/java/io/github/crabzilla/core/example1/customer/CustomerFunctions.java
package io.github.crabzilla.core.example1.customer;

import io.github.crabzilla.core.entity.EntityCommand;
import io.github.crabzilla.core.entity.EntityUnitOfWork;
import io.github.crabzilla.core.entity.Snapshot;
import io.github.crabzilla.core.entity.StateTransitionsTrackerFactory;
import io.github.crabzilla.core.example1.util.AbstractCommandsHandlerFn;
import io.github.crabzilla.core.example1.util.AbstractStateTransitionFn;
import lombok.val;

import java.util.Collections;
import java.util.List;
import java.util.function.Function;

import static io.github.crabzilla.core.entity.EntityUnitOfWork.unitOfWork;
import static io.github.crabzilla.core.example1.customer.CustomerData.*;

public class CustomerFunctions {

  // tag::StateTransitionFn[]

  public static class StateTransitionFn extends AbstractStateTransitionFn<Customer> {

    public Customer on(final CustomerCreated event, final Customer instance) {
      return instance.withId(event.getId()).withName(event.getName());
    }

    public Customer on(final CustomerActivated event, final Customer instance) {
      return instance.withActive(true).withReason(event.getReason());
    }

    public Customer on(final CustomerDeactivated event, final Customer instance) {
      return instance.withActive(false).withReason(event.getReason());
    }

  }

  // end::StateTransitionFn[]

  // tag::CommandValidatorFn[]

  public static class CommandValidatorFn implements Function<EntityCommand, List<String>> {

    @Override
    public List<String> apply(EntityCommand entityCommand) {
      return Collections.emptyList(); // all commands are valid
    }

  }

  // end::CommandValidatorFn[]

  // tag::CommandHandlerFn[]

  public static class CommandHandlerFn extends AbstractCommandsHandlerFn<Customer> {

    final StateTransitionsTrackerFactory<Customer> trackerFactory;

    public CommandHandlerFn(StateTransitionsTrackerFactory<Customer> trackerFactory) {
      this.trackerFactory = trackerFactory;
    }

    public EntityUnitOfWork handle(CreateCustomer cmd, Snapshot<Customer> snapshot) {
        return unitOfWork(cmd, snapshot.nextVersion(), snapshot.getInstance()
                .create(cmd.getTargetId(), cmd.getName()));
    }

    public EntityUnitOfWork handle(ActivateCustomer cmd, Snapshot<Customer> snapshot) {
      return unitOfWork(cmd, snapshot.nextVersion(),
              snapshot.getInstance().activate(cmd.getReason()));
    }

    public EntityUnitOfWork handle(DeactivateCustomer cmd, Snapshot<Customer> snapshot) {
      return unitOfWork(cmd, snapshot.nextVersion(),
              snapshot.getInstance().deactivate(cmd.getReason()));
    }

    public EntityUnitOfWork handle(CreateActivateCustomer cmd, Snapshot<Customer> snapshot) {
      val tracker = trackerFactory.apply(snapshot);
      val events = tracker
              .applyEvents(customer -> customer.create(cmd.getTargetId(), cmd.getName()))
              .applyEvents(customer -> customer.activate(cmd.getReason()))
              .collectEvents();
      return unitOfWork(cmd, snapshot.nextVersion(), events);
    }

  }

  // end::CommandHandlerFn[]

}

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-core/src/main/java/io/github/crabzilla/example1/customer/CustomerFunctions.kt
package io.github.crabzilla.example1.customer

import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.entity.EntityCommand
import io.github.crabzilla.core.entity.EntityCommandResult
import io.github.crabzilla.core.entity.Snapshot
import io.github.crabzilla.core.entity.StateTransitionsTrackerFactory
import io.github.crabzilla.core.exceptions.UnknownCommandException
import io.github.crabzilla.example1.resultOf
import io.github.crabzilla.example1.uowOf
import java.util.function.BiFunction
import java.util.function.Function

// tag::StateTransitionFn[]

class StateTransitionFn : BiFunction<DomainEvent, Customer, Customer> {
  override fun apply(event: DomainEvent, customer: Customer): Customer {
    return when(event) {
      is CustomerCreated -> customer.copy(_id = event.id, name =  event.name)
      is CustomerActivated -> customer.copy(isActive = true, reason = event.reason)
      is CustomerDeactivated -> customer.copy(isActive = false, reason = event.reason)
      else -> customer
    }
  }
}

// end::StateTransitionFn[]

// tag::CommandValidatorFn[]

class CommandValidatorFn : Function<EntityCommand, List<String>> { // <3>
  override fun apply(command: EntityCommand): List<String> {
    return when(command) {
      is CreateCustomer ->
        if (command.name.equals("a bad name"))
          listOf("Invalid name: ${command.name}") else listOf()
      else -> listOf() // all other commands are valid
    }
  }
}

// end::CommandValidatorFn[]

// tag::CommandHandlerFn[]

class CommandHandlerFn(private val trackerFactory: StateTransitionsTrackerFactory<Customer>) :
        BiFunction<EntityCommand, Snapshot<Customer>, EntityCommandResult> {

  override fun apply(cmd: EntityCommand, snapshot: Snapshot<Customer>): EntityCommandResult {

    val customer = snapshot.instance
    val newVersion = snapshot.version.nextVersion()

    return resultOf {
      when (cmd) {
        is CreateCustomer ->
          uowOf(cmd, customer.create(cmd.targetId as CustomerId, cmd.name), newVersion)
        is ActivateCustomer -> uowOf(cmd, customer.activate(cmd.reason), newVersion)
        is DeactivateCustomer -> uowOf(cmd, customer.deactivate(cmd.reason), newVersion)
        is CreateActivateCustomer -> {
          val tracker = trackerFactory.apply(snapshot)
          val events = tracker
                  .applyEvents({ c -> c.create(cmd.targetId as CustomerId, cmd.name) })
                  .applyEvents({ c -> c.activate(cmd.reason) })
                  .collectEvents()
          uowOf(cmd, events, newVersion)
        }
        else -> throw UnknownCommandException("for command ${cmd.javaClass.simpleName}")
      }
    }
  }
}

// end::CommandHandlerFn[]